executor/aggfuncs/func_count.go

```diff
@@ -0,0 +1,61 @@
+package aggfuncs
+
+import (
+	"github.com/juju/errors"
+
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+type baseCount struct {
+	baseAggFunc
+}
+
+type partialResult4Count struct {
+	count int64
+}
+
+func (e *baseCount) AllocPartialResult() PartialResult {
+	return PartialResult(&partialResult4Count{})
+}
+
+func (e *baseCount) ResetPartialResult(pr PartialResult) {
+	p := (*partialResult4Count)(pr)
+	p.count = 0
+}
+
+func (e *baseCount) AppendFinalResult2Chunk(sctx sessionctx.Context, pr PartialResult, chk *chunk.Chunk) error {
+	p := (*partialResult4Count)(pr)
+	chk.AppendInt64(e.ordinal, p.count)
+	return nil
+}
+
+type countOriginal struct {
+	baseCount
+}
+
+func (e *countOriginal) UpdatePartialResult(sctx sessionctx.Context, rowsInGroup []chunk.Row, pr PartialResult) error {
+	p := (*partialResult4Count)(pr)
+	p.count += int64(len(rowsInGroup))
```

Comment:
We should ignore the `NULL` values in a group, only count the non-NULL values.w

Reply:
Got.

Reply:
As far as I Know: while in `countOriginal`, the `e.args`'s len is 0, and I can't do the ` e.args[0].EvalInt(sctx, row)` in range for `rowInGroup`.

Reply:
@Xuanwo 
1. `COUNT` is special, they can have many parameters
2. You can only call `EvalInt` when the type of input parameter is integer.

Reply:
I'm confused by the `expression.Expression`.

1. The comment said that: `args stores the input arguments for an aggregate function`. And where the `input arguments` from? From the user's sql?
2. In the `AVG` aggfuncs, I found the use of `e.args[0]` and `e.args[1]`. It looks like `e.args[0]` represents `count` and `e.args[1]`represents `sum`. When and where did these value determined?

I tried to trace the calling tree but failed to answer these qusetions.

Reply:
We execute the aggregate function inside an aggregate operator, the `input arguments` are from the child of that aggregate operator. For example:
```sql
TiDB(localhost:4000) > explain select count(distinct a, b) from t;
+----------------------+------+----------------------------------------------+-------+
| id                   | task | operator info                                | count |
+----------------------+------+----------------------------------------------+-------+
| StreamAgg_8          | root | funcs:count(distinct test.t.a, test.t.b)     | 1.00  |
| └─TableReader_15     | root | data:TableScan_14                            | 3.00  |
|   └─TableScan_14     | cop  | table:t, range:[-inf,+inf], keep order:false | 3.00  |
+----------------------+------+----------------------------------------------+-------+
3 rows in set (0.00 sec)
```
In this case, we execute the `count(distinct a, b)` in the aggregate operator `StreamAgg_8`, and its input arguments come from the child operator, namely `TableReader_15`.

"In the AVG aggfuncs, I found the use of e.args[0] and e.args[1]. It looks like e.args[0] represents count and e.args[1]represents sum", this kind of `AVG` function is handle the partial result of another aggregate operator, for example:
```sql
TiDB(localhost:4000) > explain select avg(a) from t;
+------------------------+------+----------------------------------------------+-------+
| id                     | task | operator info                                | count |
+------------------------+------+----------------------------------------------+-------+
| StreamAgg_16           | root | funcs:avg(col_0, col_1)                      | 1.00  |
| └─TableReader_17       | root | data:StreamAgg_8                             | 1.00  |
|   └─StreamAgg_8        | cop  | funcs:avg(test.t.a)                          | 1.00  |
|     └─TableScan_15     | cop  | table:t, range:[-inf,+inf], keep order:false | 3.00  |
+------------------------+------+----------------------------------------------+-------+
4 rows in set (0.00 sec)
```

Here we have two aggregate operators, `StreamAgg_8` belongs to a coprocessor task and is executed on the TiKV server, `StreamAgg_16` belongs to a root task and is executed on the TiDB server.

`StreamAgg_8` handles the original data of table `t` produces partial result, which is further read by `TableReader_17` and `StreamAgg_16`. In the partial result of `StreamAgg_8`, `count` comes first and `sum` comes second. This is determined by the planner, see [this](https://github.com/pingcap/tidb/blob/master/plan/task.go#L380) for more detail.

`StreamAgg_16` handles the partial result of `StreamAgg_8`, reads `count` in the first parameter and `count` in the second parameter.

Reply:
Thank you for the wonderful explain!